clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AbstractAvroStoreClient.java

```diff
@@ -657,59 +634,96 @@ public void compute(
+      }
+
+      @Override
+      public void onRemoteComputeStateChange(boolean enabled) {
+        remoteComputationAllowed.set(enabled);
+      }
+
+      @Override
+      public void onCompletion(Optional<Exception> exception) {
+        callback.onCompletion(exception);
+      }
+    };
+
+    TransportClientStreamingCallback decoder = new ClientComputeRecordStreamDecoder<K, GenericRecord>(
+        keyList,
+        clientComputeCallback,
+        callback,
+        getDeserializationExecutor(),
+        streamingFooterRecordDeserializer,
+        () -> getComputeResultRecordDeserializer(resultSchema),
+        schemaId -> (RecordDeserializer<GenericRecord>) getDataRecordDeserializer(schemaId),
+        this::decompressRecord);
+
+    if (clientConfig.isRemoteComputationOnly() || remoteComputationAllowed.get()) {
+      compute(computeRequest, keyList, decoder, stats);
+    } else {
+      /** Multi-get fallback is on until the router tells otherwise via {@link HttpConstants.VENICE_CLIENT_COMPUTE} */
+      streamingBatchGet(keyList, decoder, stats);
+    }
+  }
+
+  private void compute(
+      ComputeRequestWrapper computeRequest,
+      List<K> keyList,
+      TransportClientStreamingCallback callback,
+      Optional<ClientStats> stats) throws VeniceClientException {
+    Map<String, String> headers = new HashMap<>(
+        computeRequest.getComputeRequestVersion() == COMPUTE_REQUEST_VERSION_V2
             ? COMPUTE_HEADER_MAP_FOR_STREAMING_V2
             : COMPUTE_HEADER_MAP_FOR_STREAMING_V3);
-    headerMap.put(VENICE_KEY_COUNT, Integer.toString(keyList.size()));
-    headerMap.put(VENICE_COMPUTE_VALUE_SCHEMA_ID, Integer.toString(schemaId));
-
-    RecordDeserializer<GenericRecord> computeResultRecordDeserializer =
-        getComputeResultRecordDeserializer(resultSchema);
-
-    transportClient.streamPost(
-        getComputeRequestPath(),
-        headerMap,
-        serializedComputeRequest,
-        new StoreClientStreamingCallback<>(keyList, callback, envelopeSchemaId -> {
-          validateComputeResponseSchemaId(envelopeSchemaId);
-          return new ComputeResponseRecordV1ChunkedDeserializer();
-        },
-            // Compute doesn't support compression
-            (envelope, compressionStrategy) -> {
-              if (!envelope.value.hasRemaining()) {
-                // Safeguard to handle empty value, which indicates non-existing key.
-                return null;
-              }
-              return new ComputeGenericRecord(
-                  computeResultRecordDeserializer.deserialize(envelope.value),
-                  computeRequestWrapper.getValueSchema());
-            },
-            envelope -> envelope.keyIndex,
-            envelope -> streamingFooterRecordDeserializer.deserialize(envelope.value)),
-        keyList.size());
+    int schemaId = getSchemaReader().getValueSchemaId(computeRequest.getValueSchema());
```

Comment:
Thank you for the suggestion! This is original compute logic. I prefer to address this optimizations in a separate PR, since this one is already substantial.